private/reducer.rkt

```diff
@@ -69,61 +71,60 @@
+  (make-reducer #:starter (fn #/variant #:consume #/init-state-maker)
+                #:consumer (fn s v #/variant #:consume #/consumer s v)
                 #:finisher finisher
                 #:early-finisher values
                 #:name name))
 
 (define into-sum (make-fold-reducer + 0 #:name 'into-sum))
 (define into-product (make-fold-reducer * 1 #:name 'into-product))
-(define into-count (make-fold-reducer (λ (s _) (add1 s)) 0 #:name 'into-count))
+(define into-count (make-fold-reducer (fn s v #/add1 s) 0 #:name 'into-count))
 
 (define (reduce red . vs)
   (define starter (reducer-starter red))
   (define consumer (reducer-consumer red))
   (define finisher (reducer-finisher red))
   (define early-finisher (reducer-early-finisher red))
-  (let loop ([tagged-state (starter)] [vs vs])
+  (w-loop loop tagged-state (starter) vs vs
     (define state (variant-value tagged-state))
-    (cond
-      [(variant-tagged-as? tagged-state '#:early-finish)
-       (early-finisher state)]
-      [(empty? vs) (finisher state)]
-      [else (loop (consumer state (first vs)) (rest vs))])))
+    (if (variant-tagged-as? tagged-state '#:early-finish)
+      (early-finisher state)
+    #/expect vs (cons v vs) (finisher state)
+    #/loop (consumer state v) vs)))
 
 (define (reduce-all red seq)
   (define starter (reducer-starter red))
   (define consumer (reducer-consumer red))
   (define finisher (reducer-finisher red))
   (define early-finisher (reducer-early-finisher red))
   (define-values (first-vs generate-rest) (sequence-generate* seq))
-  (let loop ([tagged-state (starter)]
-             [sequence-position 0]
-             [first-vs first-vs]
-             [generate-rest generate-rest])
+  (w-loop loop
+    tagged-state (starter)
+    sequence-position 0
+    first-vs first-vs
+    generate-rest generate-rest
+    
     (define state (variant-value tagged-state))
-    (cond
-      [(variant-tagged-as? tagged-state '#:early-finish)
-       (early-finisher state)]
-      [(false? first-vs)
-       (finisher state)]
-      [(not (equal? (length first-vs) 1))
-       (apply raise-result-arity-error
-              'reduce-all 1
-              (format "\n  in: sequence elements at position ~a"
-                      sequence-position)
-              first-vs)]
-      [else
-       (define-values (next-vs next-generate-rest) (generate-rest))
-       (loop (consumer state (first first-vs))
-             (add1 sequence-position)
-             next-vs
-             next-generate-rest)])))
+    (if (variant-tagged-as? tagged-state '#:early-finish)
+      (early-finisher state)
+    #/mat first-vs #f (finisher state)
+    #/expect first-vs (list first-v)
+      (apply raise-result-arity-error
+            'reduce-all 1
+            (format "\n  in: sequence elements at position ~a"
+                    sequence-position)
+            first-vs)
+    #/let ()
+      (define-values (next-vs next-generate-rest) (generate-rest))
```

Comment:
Why does this need to be in a `let`?

Reply:
This `let` is the success branch of the `expect` form. It can only be a single expression, but by using `#/let ()`, it's easy to integrate occasional blocks of local definitions into the code anyway. If you prefer, this one could be turned into `let-values`.

In [this comment](https://github.com/jackfirth/rebellion/pull/181#discussion_r314639241) I go into some more of the reasons I've treated local definition forms as afterthought and prefer to have them be errors unless explicitly introduced with a `(let () ...)`.